fvm/transactionVerifier.go

```diff
@@ -14,6 +16,123 @@ import (
 	"github.com/onflow/flow-go/module/trace"
 )
 
+type signatureType struct {
+	message []byte
+
+	errorBuilder func(flow.TransactionSignature, error) errors.CodedError
+
+	aggregateWeights map[flow.Address]int
+}
+
+type signatureEntry struct {
+	flow.TransactionSignature
+
+	signatureType
+
+	accountKey flow.AccountPublicKey
+
+	verifyErr errors.CodedError
+}
+
+func (entry *signatureEntry) newError(err error) errors.CodedError {
+	return entry.errorBuilder(entry.TransactionSignature, err)
+}
+
+func (entry *signatureEntry) matches(
+	proposalKey flow.ProposalKey,
+) bool {
+	return entry.Address == proposalKey.Address &&
+		entry.KeyIndex == proposalKey.KeyIndex
+}
+
+func (entry *signatureEntry) verify() errors.CodedError {
+	valid, err := crypto.VerifySignatureFromTransaction(
+		entry.Signature,
+		entry.message,
+		entry.accountKey.PublicKey,
+		entry.accountKey.HashAlgo,
+	)
+	if err != nil {
+		entry.verifyErr = entry.newError(err)
+	} else if !valid {
+		entry.verifyErr = entry.newError(fmt.Errorf("signature is not valid"))
+	}
+
+	return entry.verifyErr
+}
+
+func newSignatureEntries(
+	payloadSignatures []flow.TransactionSignature,
+	payloadMessage []byte,
+	envelopeSignatures []flow.TransactionSignature,
+	envelopeMessage []byte,
+) (
+	[]*signatureEntry,
+	map[flow.Address]int,
+	map[flow.Address]int,
+	error,
+) {
+	payloadWeights := make(map[flow.Address]int, len(payloadSignatures))
+	envelopeWeights := make(map[flow.Address]int, len(envelopeSignatures))
+
+	type pair struct {
+		signatureType
+		signatures []flow.TransactionSignature
+	}
+
+	list := []pair{
+		{
+			signatureType{
+				payloadMessage,
+				errors.NewInvalidPayloadSignatureError,
+				payloadWeights,
+			},
+			payloadSignatures,
+		},
+		{
+			signatureType{
+				envelopeMessage,
+				errors.NewInvalidEnvelopeSignatureError,
+				envelopeWeights,
+			},
+			envelopeSignatures,
+		},
+	}
+
+	numSignatures := len(payloadSignatures) + len(envelopeSignatures)
+	signatures := make([]*signatureEntry, 0, numSignatures)
+
+	type uniqueKey struct {
+		address flow.Address
+		index   uint64
+	}
+	duplicate := make(map[uniqueKey]struct{}, numSignatures)
+
+	for _, group := range list {
+		for _, signature := range group.signatures {
+			entry := &signatureEntry{
+				TransactionSignature: signature,
+				signatureType:        group.signatureType,
```

Comment:
Why constructing an entry structure without `accountKey`?

Would this create a possibility of hitting nil panic when reading `entry.accountKey.PublicKey` in `verify()` method?

Would it be useful to create a `signatureEntryData` with these two fields only?

Reply:
This is intentional.  The struct is used to keep track of continuation states.   accountKey and verifyErr are set by `getAccountKeys` and `verifyAccountSignatures` respectively.  These method are not accessible outside of the `verifyTransaction()`, so there's no risk of uninitialized fields.

I can add some comment to the struct and an extra nil check for clarity